### PR TITLE
Fix bugs when reading qdp file to allow one "NO" to start a new sector

### PR DESCRIPTION
**Pull request by lmytime**
### Description

This pull request is to address the bug when reading `.qdp` file to allow one "NO" to start a new sector and allow any space-like characters ahead of the "NO".

For example, in the current version `astropy.table.Table` can not read well this:
```
READ Terr 1 2
  9.0150001E+00  1.5000167E-02 -1.5000167E-02 -2.7721735E-03  5.7468953E-02 -5.7468953E-02  8.2039691E-02  7.8058078E-02
  9.0500002E+00  1.9999990E-02 -1.9999990E-02  1.1883142E-01  7.9723340E-02 -7.9723340E-02  8.3517993E-02  8.5578051E-02
NO
  1.9920001E+01  2.9999205E-02 -2.9999205E-02  2.1270609E-01  8.9663795E-02 -8.9663795E-02  9.9843595E-02  1.4176735E-01
  1.9975000E+01  2.5000445E-02 -2.5000445E-02  1.1424498E-01  8.8841605E-02 -8.8841605E-02  1.2776274E-01  1.6331280E-01
```
This file is generated by the [SPEX](https://www.sron.nl/astrophysics-spex) software and satisfies the `qdp` file standard. It should be read by the `.qdp` formatting. Thanks for @liuguanfu1120 pointing out this and testing.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: docs/changes/io.ascii/qdb-change.rst
Comment (pllim) [inline docs/changes/io.ascii/qdb-change.rst:1]: This is not the correct file name for change log fragment. Please refer to https://github.com/astropy/astropy/blob/main/docs/changes/README.rst

**Review by pllim — COMMENTED** [review]
Thanks!

Probably need a new regression test?

**Review by matteobachetti — CHANGES_REQUESTED** [review]
@lmytime thanks for your contribution! 
One small request: could you add a line to the doctest or the tests in `astropy/io/ascii/tests/test_qdp.py` to show that your modification works?

**Comment by matteobachetti** [comment]
I had a doubt that the QDP standard wanted at least two NOs to separate different groups, but the [QDP documentation](https://heasarc.gsfc.nasa.gov/ftools/others/qdp/qdp.pdf) seems to confirm that the important thing is the NO in the x coordinate. (Also tried [this tool](https://wwwastro.msfc.nasa.gov/cgi-aft/plt/htmlqdp) and the single "NO" indeed works.)

**Comment by lmytime** [comment]
Thanks. I will follow the instructions when I have time (weeks later

**Comment by lmytime** [comment]
> @lmytime thanks for your contribution! One small request: could you add a line to the doctest or the tests in `astropy/io/ascii/tests/test_qdp.py` to show that your modification works?

Test added.

**Comment by lmytime** [comment]
> I had a doubt that the QDP standard wanted at least two NOs to separate different groups, but the [QDP documentation](https://heasarc.gsfc.nasa.gov/ftools/others/qdp/qdp.pdf) seems to confirm that the important thing is the NO in the x coordinate. (Also tried [this tool](https://wwwastro.msfc.nasa.gov/cgi-aft/plt/htmlqdp) and the single "NO" indeed works.)

Yes! This is exactly what we find. Therefore, we allow astropy to read the `.qdp` file in such format.

**Comment by matteobachetti** [comment]
@lmytime thanks! The pre-commit tests complain that the code is not formatted correctly. I think this is because of the single new line before your test, and the missing newline at the end. To be sure, please run the `black` formatter on the code, and verify that the `tox -e codestyle` passes, according to the coding conventions https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions

**Comment by lmytime** [comment]
> @lmytime thanks! The pre-commit tests complain that the code is not formatted correctly. I think this is because of the single new line before your test, and the missing newline at the end. To be sure, please run the `black` formatter on the code, and verify that the `tox -e codestyle` passes, according to the coding conventions https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions

Thanks for this instruction! Done!

**Review by matteobachetti — APPROVED** [review]

File: astropy/io/ascii/tests/test_qdp.py
Comment (pllim) [inline astropy/io/ascii/tests/test_qdp.py:276]: Why not use `fp.write` here?

Comment (pllim) [reply astropy/io/ascii/tests/test_qdp.py:276]: Or simply https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text ?

File: astropy/io/ascii/tests/test_qdp.py
Comment (pllim) [inline astropy/io/ascii/tests/test_qdp.py:278]: Is it safer to test the user API (like L245 above) rather than calling a hidden function directly?

File: docs/changes/io.ascii/15392.bugfix.rst
Comment (pllim) [inline docs/changes/io.ascii/15392.bugfix.rst:1]: I do not see a test for "space-like characters head of the 'NO'" in your diff. Did I miss something?

File: astropy/io/ascii/tests/test_qdp.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_qdp.py:271]:
```suggestion
@pytest.mark.parametrize("spacelike", ["", " ", "   ", "\t"])
def test_read_qdpfile_with_oneNO_start_new_sector(tmp_path, spacelike):
    example_qdp = """
READ Terr 1 2
9.0150001E+00  1.5000167E-02 -1.5000167E-02 -2.7721735E-03  5.7468953E-02 -5.7468953E-02  8.2039691E-02  7.8058078E-02
9.0500002E+00  1.9999990E-02 -1.9999990E-02  1.1883142E-01  7.9723340E-02 -7.9723340E-02  8.3517993E-02  8.5578051E-02
{spacelike}NO
1.9920001E+01  2.9999205E-02 -2.9999205E-02  2.1270609E-01  8.9663795E-02 -8.9663795E-02  9.9843595E-02  1.4176735E-01
1.9975000E+01  2.5000445E-02 -2.5000445E-02  1.1424498E-01  8.8841605E-02 -8.8841605E-02  1.2776274E-01  1.6331280E-01
"""
```
I think the current example does not test the functionality that _any_ spacelike sequence ahead of the `NO` is accepted; this should do it (extend as needed if I've forgotten any spacelikes).

File: docs/changes/io.ascii/15392.bugfix.rst
Comment (dhomeier) [reply docs/changes/io.ascii/15392.bugfix.rst:1]: It's already testing with four spaces, but nothing else (see suggestion).

**Comment by dhomeier** [comment]
With `\n` this will accept any number of empty lines between groups; if that is allowed by the standard, this should be good now, but I could not figure it out from the docs.

**Comment by lmytime** [comment]
> With `\n` this will accept any number of empty lines between groups; if that is allowed by the standard, this should be good now, but I could not figure it out from the docs.

Ok

> With `\n` this will accept any number of empty lines between groups; if that is allowed by the standard, this should be good now, but I could not figure it out from the docs.

Yeah. I think `.qdp` file can have multiple empty lines, at least which is legal in [this](https://wwwastro.msfc.nasa.gov/qdp/wqdpxp.html). But I think it should not be implemented in this way. I will delete the `\n` character.

File: astropy/io/ascii/tests/test_qdp.py
Comment (pllim) [inline astropy/io/ascii/tests/test_qdp.py:277]:
```suggestion
    path.write_text(example_qdp)
```

**Comment by dhomeier** [comment]
> Yeah. I think `.qdp` file can have multiple empty lines, at least which is legal in [this](https://wwwastro.msfc.nasa.gov/qdp/wqdpxp.html). But I think it should not be implemented in this way. I will delete the `\n` character.

Well, it is accepted (matching `\s`), then it should be tested as well. If you don't want to accept extra lines, I'd rather change the re to `[ \t\f\v]*` (actually not sure if `\f` or `\v` are appropriate here).

**Comment by dhomeier** [comment]
I think currently the default `sep` will accept constructs like
```
   NO

    NO NO
 NO
```
This is parsed by the web applet as well, but that even takes inputs like ` N ON ONO`, so it probably only makes a weak check of the group separators.
However there is probably no harm in letting this through.
@matteobachetti could you perhaps have another look if it would make sense to honour the `delimiter` kwarg here as well, e.g.
```python
    sep = delimiter
    if delimiter is None:
        sep = r"\s"
    _new_re = rf"{sep}*NO({sep}+NO)+"
    _data_re = rf"({_decimal_re}|NO|[-+]?nan)({sep}+({_decimal_re}|NO|[-+]?nan))*)"
```
?

**Comment by lmytime** [comment]
> I think currently the default `sep` will accept constructs like
> 
> ```
>    NO
> 
>     NO NO
>  NO
> ```
> 
> This is parsed by the web applet as well, but that even takes inputs like ` N ON ONO`, so it probably only makes a weak check of the group separators. However there is probably no harm in letting this through. @matteobachetti could you perhaps have another look if it would make sense to honour the `delimiter` kwarg here as well, e.g.
> 
> ```python
>     sep = delimiter
>     if delimiter is None:
>         sep = r"\s"
>     _new_re = rf"{sep}*NO({sep}+NO)+"
>     _data_re = rf"({_decimal_re}|NO|[-+]?nan)({sep}+({_decimal_re}|NO|[-+]?nan))*)"
> ```
> 
> ?

I think the `.qdp` file standard has not been quantified well. Here, what I want to do is just to allow astropy to read `.qdp` file generated by SPEX, which uses one NO to start a new sector. Therefore, we can have this new function.
I think there is no weird `.qdp` file... after all, they are generated by good software and good people. ;)
The reason why I use `\s` is for robustness. Also, the meaning is immediate when seeing \s.

**Comment by dhomeier** [comment]
> generated by SPEX, which uses one NO to start a new sector. Therefore, we can have this new function.
> I think there is no weird `.qdp` file... after all, they are generated by good software and good people. ;)

That assumption of course is a recipe for disaster, generally speaking, when designing robust software. ;-)
But I agree that the worst that's likely to happen is that some people end up working with data from a broken frontend app or provider for some time without noticing.

> The reason why I use `\s` is for robustness. Also, the meaning is immediate when seeing \s.

Yes, that's possibly also the reason it was chose as default separator in #11256.
But there it was also decided not to accept inputs like `N ON ON   \t O`.
Since I don't know more about this format either, I am leaving it up to you and @matteobachetti as people who are using it, to decide if you are happy with this version.

**Comment by pllim** [comment]
I agree that assuming people will always make good software is not a safe thing to do. If you are aware of the limitation but do not want to fix it, please at least clearly document it as a known limitation for the future. Thanks!

**Comment by matteobachetti** [comment]
@dhomeier the previous version worked perfectly for the use cases I was aware of (Xspec output, Swift light curves). @lmytime found a slightly different format, that can be easily accommodated. I personally wouldn't go imagining  more complicated scenarios, with a format that is so poorly documented. Every new thing we allow potentially introduces bugs in other plausible use cases. 

For example: what about a line that starts with "NO" and then has valid data? This might be a point of contention, because the documentation says that a new dataset should start, but it's perfectly possible that some random user has invalid data in the x axis, and I would rather allow them having a single table with an invalid data point.

Personally, I wouldn't even allow for white space at the start of the line 😅 

One thing I do agree with @dhomeier is that the format specification does allow for other separators, and we should make sure this new addition doesn't clash with that use case. It is easy to do another test equal to the one @lmytime added, but substituting white spaces with a comma and specifying the correct separator.

**Comment by dhomeier** [comment]
Yes, I did not argue for allowing more styles; if anything, rather for being a bit more restrictive.

> One thing I do agree with @dhomeier is that the format specification does allow for other separators, and we should make sure this new addition doesn't clash with that use case. It is easy to do another test equal to the one @lmytime added, but substituting white spaces with a comma and specifying the correct separator.

So it would make sense, with `delimiter=','` set,  to accept `„NO` but not `    NO`?

**Comment by matteobachetti** [comment]
No, that would be against the format specification! NO has to be the first datum in the line.
I see what you mean though: if the separator is a space, starting with a space would imply missing data. Right? I agree that allowing spaces is probably error prone in this sense

**Comment by dhomeier** [comment]
Or, if `,` is the separator, NO could still be considered the first datum in `,NO`. But I agree, really this would constitute an empty datum, so that case is a bit inconsistent with the blank separator. But spaces can certainly slip in more easily.

**Comment by dhomeier** [comment]
One thing that might yet be added, though not directly related to this issue, is documentation for the `delimiter` parameter. Could also be worth noting that it only applies to data, not to command lines.

**Comment by matteobachetti** [comment]
Why don't we just, for now, prohibit starting spaces in lines starting with NO? Usually, qdp files spit out by our favorite programs do not write spaces, and we all agree that they can confuse the notation. 
This simplifies the code, and still the code works for all the use cases we know of. We can have either @lmytime write two more lines of documentation to document the separator, or open an issue and I will take care of it (since it's really a different issue).
If we limit the new modification to this simple update, I think it's also safe to backport, what do you think @saimn?

**Comment by matteobachetti** [comment]
@lmytime would you like to finish this up?

**Comment by lmytime** [comment]
> @lmytime would you like to finish this up?

I will. But recently I have been busy. Maybe I will try to finish this in one month.

**Comment by mhvk** [comment]
@lmytime - would you have time to finish up this PR?

**Comment by pllim** [comment]
Does any maintainer want to take over? Looks like it got closed by bot.

**Comment by neutrinoceros** [comment]
(not a maintainer but feel free to throw it my way)

**Comment by dhomeier** [comment]
Last we've been waiting for a reply from @lmytime, but 5 months have passed since, so we should not wait much longer.
QDP is a bit special, and not really strictly defined, but I think making sure that [Matteo's suggestions](https://github.com/astropy/astropy/pull/15392#issuecomment-1796428991) are followed, this should be about ready.

**Comment by neutrinoceros** [comment]
@dhomeier just let me know your preference, I'm fine either way !

**Pull request closed without merging**
